parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java

```diff
@@ -0,0 +1,330 @@
+      // bloom filter is based on hash and cannot eliminate based on lt or ltEq or gt or gtEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean gt(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on lt or ltEq or gt or gtEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean gtEq(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on lt or ltEq or gt or gtEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean eq(BoundReference<T> ref, Literal<T> lit) {
+      int id = ref.fieldId();
+      ColumnDescriptor col = cols.get(id);
+      if (col == null) {
+        return ROWS_MIGHT_MATCH;
+      } else {
+        Boolean hasNoBloomFilter = fieldsWithNoBloom.contains(id);
+        if (hasNoBloomFilter) {
+          return ROWS_MIGHT_MATCH;
+        }
+        BloomFilter bloom = getBloomById(id);
+        Type type = types.get(id);
+        T value = lit.value();
+        return shouldRead(col, value, bloom, type);
+      }
+    }
+
+    private BloomFilter getBloomById(int id) {
+      if (bloomCache == null) {
+        bloomCache = Maps.newHashMap();
+      }
+      if (bloomCache.containsKey(id)) {
+        return bloomCache.get(id);
+      } else {
+        ColumnChunkMetaData columnChunkMetaData = columnMetaMap.get(id);
+        BloomFilter bloomFilter = bloomReader.readBloomFilter(columnChunkMetaData);
+        if (bloomFilter == null) {
+          throw new IllegalStateException("Failed to read required bloom filter for id: " + id);
+        } else {
+          bloomCache.put(id, bloomFilter);
+        }
+
+        return bloomFilter;
+      }
+    }
+
+    private <T> boolean shouldRead(ColumnDescriptor col, T value, BloomFilter bloom, Type type) {
+      long hashValue = 0;
+      switch (type.typeId()) {
+        case INTEGER:
+        case LONG:
+        case FLOAT:
+        case DOUBLE:
+        case DATE:
+        case TIME:
+        case TIMESTAMP:
+          hashValue = bloom.hash(value);
+          return bloom.findHash(hashValue) ? ROWS_MIGHT_MATCH : ROWS_CANNOT_MATCH;
+        case STRING:
+          hashValue = bloom.hash(Binary.fromCharSequence((CharSequence) value));
+          return bloom.findHash(hashValue) ? ROWS_MIGHT_MATCH : ROWS_CANNOT_MATCH;
+        case BINARY:
+        case FIXED:
+          hashValue = bloom.hash(Binary.fromReusedByteBuffer((ByteBuffer) value));
+          return bloom.findHash(hashValue) ? ROWS_MIGHT_MATCH : ROWS_CANNOT_MATCH;
+        case DECIMAL:
+          BigDecimal decimalValue = (BigDecimal) value;
+          switch (col.getPrimitiveType().getPrimitiveTypeName()) {
+            case INT32:
+              hashValue = bloom.hash(decimalValue.unscaledValue().intValue());
+              return bloom.findHash(hashValue) ? ROWS_MIGHT_MATCH : ROWS_CANNOT_MATCH;
```

Comment:
While I appreciate the use of `ROWS_MIGHT_MATCH` and `ROWS_CANNOT_MATCH`, I think this is no different than returning the result of `findHash` directly, right?

Reply:
Right. Fixed. Thanks!